src/output/plugins/MacOSOutputPlugin.cxx

```diff
@@ -0,0 +1,553 @@
+}
+
+#ifdef ENABLE_DSD
+
+void
+MacOSOutput::SetupDop(const AudioFormat audio_format, PcmExport::Params &params)
+{
+	assert(audio_format.format == SampleFormat::DSD);
+	
+	/* pass 24 bit to Setup() */
+	
+	AudioFormat dop_format = audio_format;
+	dop_format.format = SampleFormat::S24_P32;
+	dop_format.sample_rate = params.CalcOutputSampleRate(audio_format.sample_rate);
+	
+	const AudioFormat check = dop_format;
+	
+	Setup(dop_format);
+	
+	/* if the device allows only 32 bit, shift all DoP
+	 samples left by 8 bit and leave the lower 8 bit cleared;
+	 the DSD-over-USB documentation does not specify whether
+	 this is legal, but there is anecdotical evidence that this
+	 is possible (and the only option for some devices) */
+	params.shift8 = dop_format.format == SampleFormat::S32;
+	
+	if (dop_format.format == SampleFormat::S32)
+		dop_format.format = SampleFormat::S24_P32;
+	
+	if (dop_format != check)
+	{
+		/* no bit-perfect playback, which is required
+		 for DSD over USB */
+		throw std::runtime_error("Failed to configure DSD-over-PCM, no suitable format available.");
+	}
+
+}
+
+#endif
+
+void
+MacOSOutput::SetupOrDop(AudioFormat &audio_format, PcmExport::Params &params
+#ifdef ENABLE_DSD
+					   , bool dop
+#endif
+)
+{
+#ifdef ENABLE_DSD
+	std::exception_ptr dop_error;
+	if(audio_format.format == SampleFormat::DSD){
+		if (dop)
+		{
+			try
+			{
+				params.dop = true;
+				SetupDop(audio_format, params);
+				return;
+			}
+			catch (...)
+			{
+				dop_error = std::current_exception();
+				// DoP was unsuccessful, proceed with PCM output
+				params.dop = false;
+				audio_format.format = SampleFormat::S32;
+			}
+		}
+		else
+			/** If DoP is not configured switch to
+			 *	PCM output (DSD direct not possible
+			 *	on MacOS).
+			 */
+			audio_format.format = SampleFormat::S32;
+	}
+	
+	try
+	{
+#endif
+		Setup(audio_format);
+#ifdef ENABLE_DSD
+	}
+	catch (...)
+	{
+		if (dop_error)
+		/* if DoP was attempted, prefer returning the
+		 original DoP error instead of the fallback
+		 error */
+			std::rethrow_exception(dop_error);
+		else
+			throw;
+	}
+#endif
+}
+
+void
+MacOSOutput::Open(AudioFormat &audio_format)
+{
+	PcmExport::Params params;
+	
+#ifdef ENABLE_DSD
+	bool dop = dop_setting;
+	params.dop = false;
+#endif
+	
+	try
+	{
+		SetupOrDop(audio_format, params
+#ifdef ENABLE_DSD
+				   , dop
+#endif
+				   );
+	}
+	catch (...)
+	{
+		std::throw_with_nested(FormatRuntimeError("Error opening MacOS output device \"%s\"", device_name.c_str()));
+	}
+	
+#ifdef ENABLE_DSD
+	if (params.dop)
+		FormatDebug(macos_output_domain, "DoP (DSD over PCM) enabled");
+#endif
+	
+	pcm_export->Open(audio_format.format, audio_format.channels, params);
+	/** Setup the ring_buffer to hold
+	 *	BUFFER_TIME_MS or four times
+	 *	the device frame buffer.
+	 */
+	UInt32 dev_frame_buffer = device.GetBufferSize();
+	if(dev_frame_buffer == 0)
+		throw std::runtime_error("Cannot get valid frame buffer size from device.");
```

Comment:
Leaks memory

Reply:
How so? Apart from memory allocated within the device class that should be freed when the output is closed, nothing was allocated.

Reply:
`PcmExport::Open()` may have allocated memory.

Reply:
OK, moved `PcmExport::Open()` to end of the block.